### DISCUSSION TITLE
Use `ensureQueryData` with invalidations

### DISCUSSION THREAD

**Discussion by samiBendou** [TanStack/query#6424] [Q&A] [answered]
Hello everyone,

I had to use `ensureQueryData` in order to retrieve some query data that might already be cached. The data should be fresh, so I might have to use `fetchQuery` if the query data has been invalidated (or stale btw).

To solve this issue I came with this function that have the expected behavior :

```typescript
async function ensureFreshQueryData<
  TQueryFnData,
  TError,
  TData,
  TQueryKey extends QueryKey = QueryKey
>(
  queryClient: QueryClient,
  query: WithRequired<
    FetchQueryOptions<TQueryFnData, TError, TData, TQueryKey>,
    "queryKey"
  >
): Promise<any> {
  const state = queryClient.getQueryState(query.queryKey);
  const data = queryClient.getQueryData(query.queryKey);
  if (!state || state.isInvalidated || !data) {
    return await queryClient.fetchQuery(query);
  }

  return data;
}
```

I am not sure if this is the right way to do that. I would like to know if there is something simpler to do. If not I think that would be a good idea to integrate `ensureFreshQueryData` as a features of the `queryClient`.

Thanks a lot !

**Comment by TkDodo** [comment, marked as answer]
just use `fetchQuery` and pass a `staleTime` to it.

**Reply by samiBendou** [reply]
Wow that was fast, thank you ! :+1:
